crates/zir-codegen/src/lib.rs

```diff
@@ -155,52 +371,147 @@ impl FunctionSignature {
+/// 2. Provide target configuration in `target_config()`
+/// 3. Compile MIR to native IR in `codegen_unit()`
+/// 4. Combine parallel results in `join_codegen()`
+/// 5. Produce final output in `link()`
 ///
 /// # Example Implementation
 ///
 /// ```ignore
 /// struct MyCraneliftBackend { /* ... */ }
 ///
 /// impl CodegenBackend for MyCraneliftBackend {
-///     fn compile_function<'zir>(
-///         &mut self,
-///         body: &Body<'zir>,
-///         signature: FunctionSignature,
-///     ) -> CodegenResult<()> {
-///         // Convert MIR to Cranelift IR
-///         // Add to module
-///         Ok(())
+///     fn name(&self) -> &'static str {
+///         "cranelift"
 ///     }
 ///
-///     fn compile_to_ir<'zir>(
+///     fn init(&self, sess: &Session) {
+///         // Initialize backend with session configuration
+///     }
+///
+///     fn codegen_unit<'a>(
 ///         &mut self,
-///         body: &Body<'zir>,
-///         signature: FunctionSignature,
-///     ) -> CodegenResult<IrOutput> {
-///         // Return CLIF text representation
-///         Ok(IrOutput::Text(clif_text))
+///         unit: CodegenUnit<'a>,
+///     ) -> CodegenResult<OngoingCodegen> {
+///         // Compile the unit
+///         Ok(Box::new(results))
+///     }
+///
+///     fn join_codegen(
+///         &self,
+///         ongoing: OngoingCodegen,
+///         sess: &Session,
+///         outputs: &OutputFilenames,
+///     ) -> CodegenResult<(CodegenResults, HashMap<WorkProductId, WorkProduct>)> {
+///         // Combine results
+///         Ok((results, work_products))
 ///     }
 ///
-///     fn finalize(self: Box<Self>) -> CodegenResult<CodegenResults> {
-///         // Finalize the module and return compiled code
-///         Ok(CodegenResults::default())
+///     fn link(
+///         &self,
+///         sess: &Session,
+///         results: CodegenResults,
+///         outputs: &OutputFilenames,
+///     ) -> CodegenResult<()> {
+///         // Link the final output
+///         Ok(())
 ///     }
 /// }
 /// ```
 pub trait CodegenBackend: Any {
-    /// Returns the name of this backend.
+    /// Returns the name of this backend (e.g., "cranelift", "llvm").
     fn name(&self) -> &'static str;
 
-    /// Compiles a MIR function body.
+    /// Initializes the backend with session configuration.
+    ///
+    /// This is called once before any codegen operations. Backends
+    /// can use this to set up internal state based on compiler options.
+    fn init(&self, _sess: &Session) {}
+
+    /// Returns target-specific configuration.
     ///
-    /// This method translates MIR to the backend's native IR format
-    /// and adds it to the current module being built.
+    /// This allows backends to report what features they support for
+    /// the current target, enabling frontend decisions about code generation.
+    fn target_config(&self, _sess: &Session) -> TargetConfig {
+        TargetConfig::for_host()
+    }
```

Comment:
```rust
51TargetConfig {
52            target_features: vec![],
53            unstable_target_features: vec![],
54            // `true` is used as a default so backends need to acknowledge when they do not
55            // support the float types, rather than accidentally quietly skipping all tests.
56            has_reliable_f16: true,
57            has_reliable_f16_math: true,
58            has_reliable_f128: true,
59            has_reliable_f128_math: true,
60        }
```